waveform_modes.py

```diff
@@ -315,6 +315,29 @@ def indices(self, ell_m):
             raise ValueError("Input `ell_m` should be an Nx2 sequence of integers")
         return ell_m[:, 0] * (ell_m[:, 0] + 1) - self.ell_min ** 2 + ell_m[:, 1]
 
+
+    def inner_prod(self, b):
+        """Compute the all-angles inner product <self, b>.
+        self and b must have the same times and set of (ell,m) modes.
+
+        Parameters
+        ----------
+        b : WaveformModes object
+            The other set of modes to inner product with.
+
+        Returns
+        -------
+        inner_prod : complex
+            <self, b>
+        """
+
+        # QUESTION/TODO: Should we do validation here?
+        # Probably? Ensure that times and LMs are the same, not just same shape!
```

Comment:
Perhaps leave it as an option with an argument that the user can turn on and off?

Reply:
I like this idea.

Reply:
I don't like the idea of turning it off.  The validation should be quick, so there would be no good reason to skip it.  Moreover, I expect that it would be reasonable to automatically do some interpolation to the intersection of the two timeseries, and similar things, rather than just checking.  Maybe an option to issue warnings rather than raise — for example if the spin weights are not equal, or if some LM modes are not present in one.

Reply:
Thanks for pointing out the spin as well. So validation should be default.

Different s's: fail

Different set of (ell,m)'s: could fail, or only use the set of (ell,m)'s in the intersection of the two waveforms. Should this be an option?

Different set of times: could fail, or interpolate to a common set of times on the intersection between the two waveforms. But I'm not sure what timestep should you use if the timesteps differ. Thoughts?

Reply:
> Different set of (ell,m)'s: could fail, or only use the set of (ell,m)'s in the intersection of the two waveforms. Should this be an option?

I'd say at least a warning would be appropriate, but I think it should be allowed, because IIRC the output from `WaveformModes.transform` will have a larger ell_max than the input [as it should, whenever there's a boost or (super)translation].

> But I'm not sure what timestep should you use if the timesteps differ.

The [method I created for extrapolation](https://github.com/moble/scri/blob/5875404aceb42905ad2d387278c3341d9c2e6d10/extrapolation.py#L48-L64) is not unique, but I think it's sensible, and it's been tested over time.

Reply:
For now I updated the PR to validate equality spins, LMs, and times. The argument for intersections of LMs and times is good. I propose that the default behavior is require everything matches, unless an optional parameter is passed saying "I know what I'm doing" and to use the intersections of LMs and times — because there may be interpolation which is a potentially expensive operation. Thoughts?

Reply:
Given the conversation I like the idea of having the default be the validation, but it's always good to allow the user to skip it. I don't imagine the interpolation would be too costly, but regardless its good to allow a user to skip it for a rare case.

Reply:
I definitely agree with calling this `inner_product` but then I do propose we change the name of the `inner_product` function in `mode_calculations`. I think having both of them with the same name will be too confusing.

Reply:
Yeah, it's not just a code issue but a mathematics issue: lots of vector spaces have inner products, and there might even be different inner products on the same vector space, but we rarely give them unique names — typically <a,b> and you have to know from context that the operation < , > is on VxV->C, or WxW->C, etc.

One of these is the inner product on the space of timeseries R->C (or some other V->C), and the other is the inner product on the space of S^2xR -> C.

Is it clearer if we call one timeseries_inner_product, and the other scri_inner_product?

I would say that, similar to mathematics usage, we do have context whenever one of these methods is invoked. One of them is a member of WaveformModes and takes as argument a second wm, while the other method takes two vectors.

Reply:
There are two different types of "validation" happening here, and you should never be able to *skip* either one.  The first type is checking for things that can be handled but will affect the code path — like different time steps or modes.  In this case, the option should be between a default of printing a warning but doing what needs to be done, or just doing what needs to be done.  The second type is something like checking the spin weights; the "inner product" of functions with different spin weights will be a coordinate-dependent thing, and thus pretty meaningless.  We therefore want to strongly discourage people from doing this.  In this case, the option could be between raising an error or printing a warning.  If the user doesn't like seeing the warning s/he is free to use `warnings.catch_warnings` or `contextlib.redirect_stderr`, but we want to impose that penalty because the cost of silence will be bad physics.

Reply:
Am now validating, and have introduced options allow_LM_differ and allow_times_differ. These will respectively slice (in ell, m) and/or interpolate (in time) if needed to perform the overlap.

Reply:
Note, this PR is blocked by #9 if you try to both slice in LM and interpolate in time.

Reply:
> Is it clearer if we call one timeseries_inner_product, and the other scri_inner_product?
I think this would be better

Reply:
I just rebased on top of master after #10 was merged, so this is no longer blocked by the issue. I think the only remaining issue is names for the two different methods, unless we want to consider whether scipy.integrate.simps is the correct approach or not.

Reply:
I don't expect that it would make a huge difference numerically, but I would favor replacing `simps` with a call to `InterpolatedUnivariateSpline.integral(t1, t2)`.  On that note, it seems like adding arguments named something like `t1=None, t2=None` would make sense, if people want to cut out junk radiation or ringdown, for example.  These can be passed directly to `intersection`.

Reply:
On a related note, I would have thought that the most efficient way would be to do the sum over modes first, and then the integration over time.  Especially if you switch to spline integration, splines are so slow (and annoying to call) that this would surely improve the speed dramatically.  Thoughts?

Reply:
Good point about summing over modes first. I should have thought of that.
Let me investigate `InterpolatedUnivariateSpline.integral` in a bit...

Reply:
`InterpolatedUnivariateSpline` is more annoying for the basic `inner_product`. As it is written now, mode_calculations.py:inner_product will happily integrate along some axis of an ndarray. Using `InterpolatedUnivariateSpline`, we would need to build k interpolants, where k is the product of the # of elements in the axes that are not being integrated. Maybe that function should just be restricted to operating on vectors, rather than ndarrays?

Reply:
`InterpolatedUnivariateSpline` does not seem to be too much slower than `simps`. Now using it in `WaveformModes.inner_product`, but haven't decided what to do for the more general `mode_calculations.py:inner_product`.